worlds/apquest/items.py

```diff
@@ -0,0 +1,110 @@
+
+    # Creating items should generally be done via the world's create_item method.
+    # First, we create a list containing all the items that always exist.
+
+    itempool = [
+        world.create_item("Key"),
+        world.create_item("Sword"),
+        world.create_item("Shield"),
+        world.create_item("Health Upgrade"),
+        world.create_item("Health Upgrade"),
+    ]
+
+    # Some items may only exist if the player enables certain options.
+    # In our case, If the hammer option is enabled, the sixth item is the Hammer.
+    # Otherwise, we add a filler Confetti Cannon.
+    if world.options.hammer:
+        # Once again, it is important to stress that even though the Hammer doesn't always exist,
+        # it must be present in the worlds location_name_to_id.
+        # Whether it is actually in the itempool is determined purely by whether we create and add the item here.
+        itempool.append(world.create_item("Hammer"))
+
+    # Archipelago requires that each world submits as many locations as it submits items.
+    # This is what filler is for. In our case, that's the Confetti Cannon.
+    # Creating filler items works the same as any other item. But there is a question:
+    # How many Confetti Cannons do we actually need to create?
+    # In regions.py, we created either six or seven locations depending on the "extra_starting_chest" option.
+    # In this function, we have created five or six items depending on whether the "hammer" option is enabled.
+    # We *could* have a really complicated if-else tree checking the options again, but there is a better way.
+    # We can compare the size of our itempool so far to the amount of locations in our world.
+
+    # The length of our itempool is easy to determine, since we have it as a list.
+    amount_of_items = len(itempool)
+
+    # The amount of locations is also easy to determine, but we have to be careful.
+    # Just calling len(world.get_locations()) would report an incorrect number, because of our *event locations*.
+    # What we actually want is the amount of *unfilled* locations. Luckily, there is a helper method for this:
+    amount_of_unfilled_locations = len(world.multiworld.get_unfilled_locations(world.player))
+
+    # Now, we just subtract the amount of items from the amount of locations to get the number of empty item slots.
+    needed_amount_of_filler = amount_of_unfilled_locations - amount_of_items
+
+    # Finally, we create that many filler items and add them to the itempool.
+    # Important: Make sure you're actually submitting different instances of the item!
+    # In Python, it can be easy to accidentally submit two references to the same object rather than two objects.
+    itempool += [world.create_item("Confetti Cannon") for _ in range(needed_amount_of_filler)]
+
+    # Finally, we just have append our itempool to the multiworld itempool.
+    world.multiworld.itempool += itempool
```

Comment:
I still haven't found the time to fully review all this, but pamymaf saying:
> The only thing that I had to look up how to do was push an item to starting inventory.

reminded me of one thought I had when I was scribbling my own examples.

-----

IIUC for 99% of apworlds, the `create_items()` method will eventually end in:
a) appending to `multiworld.itempool`
b) optionally, also calling `multiworld.push_precollected()`

with all other code in the method being ultimately just set up for a or b. In particular, nothing "really happens" to the multiworld until you do a or b.

If I have all of that right, I think it's worth belaboring this comment to explain that. Even if we can't fit a real `push_precollected()` use into APQuest's design.

Reply:
Added something for precollected items.